pkg/asset/installconfig/aws/session.go

```diff
@@ -60,48 +58,115 @@ func GetSession() (*session.Session, error) { return GetSessionWithOptions() }
 // and, if no creds are found, asks for them and stores them on disk in a config file
 func GetSessionWithOptions(optFuncs ...SessionOptions) (*session.Session, error) {
 	options := session.Options{
+		Config:            aws.Config{MaxRetries: aws.Int(0)},
 		SharedConfigState: session.SharedConfigEnable,
 	}
 	for _, optFunc := range optFuncs {
 		optFunc(&options)
 	}
 
+	_, err := getCredentials(options)
+	if err != nil && errCodeEquals(err, "NoCredentialProviders") {
+		if err = getUserCredentials(); err != nil {
+			return nil, err
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+
 	ssn := session.Must(session.NewSessionWithOptions(options))
+	ssn = ssn.Copy(&aws.Config{MaxRetries: aws.Int(25)})
+	ssn.Handlers.Build.PushBackNamed(request.NamedHandler{
+		Name: "openshiftInstaller.OpenshiftInstallerUserAgentHandler",
+		Fn:   request.MakeAddToUserAgentHandler("OpenShift/4.x Installer", version.Raw),
+	})
+	return ssn, nil
+}
 
+func getCredentials(options session.Options) (*credentials.Credentials, error) {
 	sharedCredentialsProvider := &credentials.SharedCredentialsProvider{}
-	ssn.Config.Credentials = credentials.NewChainCredentials([]credentials.Provider{
+	providers := []credentials.Provider{
 		&credentials.EnvProvider{},
 		sharedCredentialsProvider,
-	})
-
-	creds, err := ssn.Config.Credentials.Get()
-	if err == nil {
-		switch creds.ProviderName {
-		case sharedCredentialsProviderName:
-			onceLoggers[sharedCredentialsProviderName].Do(func() {
-				logrus.Infof("Credentials loaded from the %q profile in file %q", sharedCredentialsProvider.Profile, sharedCredentialsProvider.Filename)
-			})
-		case envProviderName:
-			onceLoggers[envProviderName].Do(func() {
-				logrus.Info("Credentials loaded from default AWS environment variables")
-			})
-		}
 	}
-	if err == credentials.ErrNoValidProvidersFoundInChain {
-		err = getCredentials()
+
+	creds := credentials.NewChainCredentials(providers)
```

Comment:
Can you explain why we need to attempt to get the credentials from the custom chain first? My understanding is that the env provider and the shared creds provider are both providers used in the default credentials chain. Is it that we are changing the priority of the providers?

Reply:
Because that is the existing behavior today, and i don't intent to change that for supporting new methods.

if there is consensus that default code path and priority makes sense, taking that on in separate follow up PR would be better.